crates/red_knot_python_semantic/src/types.rs

```diff
@@ -75,13 +75,44 @@ pub(crate) fn global_symbol_ty_by_name<'db>(db: &'db dyn Db, file: File, name: &
     symbol_ty_by_name(db, global_scope(db, file), name)
 }
 
-/// Shorthand for `symbol_ty` that looks up a symbol in the builtins.
+/// Enumeration of various core stdlib modules, for which we have dedicated Salsa queries.
 ///
-/// Returns `Unbound` if the builtins module isn't available for some reason.
+/// Things will start getting very strange during type-checking if one of these doesn't exist.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) enum CoreStdlibModule {
+    Builtins,
+    Types,
+    Typeshed,
+}
+
+impl CoreStdlibModule {
+    /// Retrieve the global scope of the given module.
+    ///
+    /// Returns `None` if the given module isn't available for some reason.
+    pub(crate) fn global_scope(self, db: &dyn Db) -> Option<ScopeId<'_>> {
+        match self {
+            Self::Builtins => builtins_scope(db),
+            Self::Types => types_scope(db),
+            Self::Typeshed => typeshed_scope(db),
+        }
```

Comment:
I'm inclined to inreverse the `CoreStdlibModule` -> `*_scope` dependency:

* add a `module_name` method 
```suggestion
    let types_file = resolve_module(db, self.module_name())?.file();
    Some(global_scope(db, types_file))
```

Reply:
I think this would mean that getting the scope of `builtins`, `types`, etc. could not be Salsa-tracked. I'm open to trying that out but would rather do it in a separate PR so we can look at how it affects performance.